In [1]:
import pandas as pd
import numpy as np
from IPython.display import display

In [2]:

datasets = {
    "hotel_bookings": "../data/raw/hotel_bookings.csv",
    "customers": "../data/raw/olist_customers_dataset.csv",
    "geolocation": "../data/raw/olist_geolocation_dataset.csv",
    "order_items": "../data/raw/olist_order_items_dataset.csv",
    "payments": "../data/raw/olist_order_payments_dataset.csv",
    "reviews": "../data/raw/olist_order_reviews_dataset.csv",
    "orders": "../data/raw/olist_orders_dataset.csv",
    "products": "../data/raw/olist_products_dataset.csv",
    "sellers": "../data/raw/olist_sellers_dataset.csv",
    "category_translation": "../data/raw/product_category_name_translation.csv"
}



dfs = {}

for name, path in datasets.items():
    dfs[name] = pd.read_csv(path)

def analyze_dataset(name, df):

    print("\n" + "="*120)
    print(f"DATASET: {name.upper()}")
    print("="*120)


    print("\n[1] SAMPLE DATA (10 ROWS)")

    sample_size = min(10, len(df))
    display(df.sample(sample_size))


    print("\n[2] DATA SHAPE")

    shape_df = pd.DataFrame({
        "Rows": [df.shape[0]],
        "Columns": [df.shape[1]]
    })

    display(shape_df)


    print("\n[3] DATA TYPES")

    dtype_df = pd.DataFrame({
        "Column": df.columns,
        "Data Type": df.dtypes.astype(str)
    })

    display(dtype_df)


    print("\n[4] DESCRIPTIVE STATISTICS")

    display(df.describe(include="all").T)

    print("\n[5] MISSING VALUES")

    missing_df = pd.DataFrame({
        "Missing Count": df.isnull().sum(),
        "Missing Percentage (%)":
            round(df.isnull().mean() * 100, 2)
    })

    missing_df = (
        missing_df
        .sort_values(
            by="Missing Percentage (%)",
            ascending=False
        )
    )

    display(missing_df)

    print("\n[6] DUPLICATED ROWS")

    duplicate_count = df.duplicated().sum()

    duplicate_df = pd.DataFrame({
        "Duplicate Rows": [duplicate_count],
        "Duplicate Percentage (%)": [
            round(
                duplicate_count / len(df) * 100,
                2
            )
        ]
    })

    display(duplicate_df)


    print("\n[7] CARDINALITY")

    cardinality_df = pd.DataFrame({
        "Unique Values": df.nunique()
    })

    cardinality_df = (
        cardinality_df
        .sort_values(
            by="Unique Values",
            ascending=False
        )
    )

    display(cardinality_df)


for name, df in dfs.items():
    analyze_dataset(name, df)


DATASET: HOTEL_BOOKINGS

[1] SAMPLE DATA (10 ROWS)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
108590,City Hotel,0,45,2017,March,13,27,1,2,2,...,No Deposit,14.0,NaN,0,Transient,88.20,0,2,Check-Out,2017-03-30
21757,Resort Hotel,0,0,2016,March,11,7,1,1,2,...,No Deposit,6.0,NaN,0,Transient,29.00,0,1,Check-Out,2016-03-09
76881,City Hotel,1,239,2015,October,42,17,0,1,2,...,Non Refund,1.0,NaN,0,Transient,60.00,0,0,Canceled,2015-08-17
117971,City Hotel,0,308,2017,August,32,7,1,5,2,...,No Deposit,9.0,NaN,0,Transient-Party,89.10,0,1,Check-Out,2017-08-13
63003,City Hotel,1,66,2017,January,4,25,2,5,1,...,No Deposit,9.0,NaN,0,Transient,87.21,0,0,Canceled,2016-12-30
74897,City Hotel,1,326,2015,September,37,8,0,1,2,...,Non Refund,1.0,NaN,0,Transient,60.00,0,0,Canceled,2015-05-28
119178,City Hotel,0,213,2017,August,35,28,1,3,1,...,No Deposit,19.0,NaN,0,Transient-Party,104.00,0,0,Check-Out,2017-09-01
113193,City Hotel,0,0,2017,June,22,3,0,1,2,...,No Deposit,NaN,NaN,0,Transient,170.00,0,1,Check-Out,2017-06-04
95861,City Hotel,0,0,2016,August,35,24,0,1,2,...,No Deposit,152.0,NaN,0,Transient,73.07,0,0,Check-Out,2016-08-25
104654,City Hotel,0,1,2017,January,3,15,2,3,1,...,No Deposit,9.0,NaN,0,Transient,100.47,0,0,Check-Out,2017-01-20



[2] DATA SHAPE


,Rows,Columns
0,119390,32



[3] DATA TYPES


,Column,Data Type
hotel,hotel,object
is_canceled,is_canceled,int64
lead_time,lead_time,int64
arrival_date_year,arrival_date_year,int64
arrival_date_month,arrival_date_month,object
arrival_date_week_number,arrival_date_week_number,int64
arrival_date_day_of_month,arrival_date_day_of_month,int64
stays_in_weekend_nights,stays_in_weekend_nights,int64
stays_in_week_nights,stays_in_week_nights,int64
adults,adults,int64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
hotel,119390,2,City Hotel,79330,NaN,NaN,NaN,NaN,NaN,NaN,NaN
is_canceled,119390.0,NaN,NaN,NaN,0.370416,0.482918,0.0,0.0,0.0,1.0,1.0
lead_time,119390.0,NaN,NaN,NaN,104.011416,106.863097,0.0,18.0,69.0,160.0,737.0
arrival_date_year,119390.0,NaN,NaN,NaN,2016.156554,0.707476,2015.0,2016.0,2016.0,2017.0,2017.0
arrival_date_month,119390,12,August,13877,NaN,NaN,NaN,NaN,NaN,NaN,NaN
arrival_date_week_number,119390.0,NaN,NaN,NaN,27.165173,13.605138,1.0,16.0,28.0,38.0,53.0
arrival_date_day_of_month,119390.0,NaN,NaN,NaN,15.798241,8.780829,1.0,8.0,16.0,23.0,31.0
stays_in_weekend_nights,119390.0,NaN,NaN,NaN,0.927599,0.998613,0.0,0.0,1.0,2.0,19.0
stays_in_week_nights,119390.0,NaN,NaN,NaN,2.500302,1.908286,0.0,1.0,2.0,3.0,50.0
adults,119390.0,NaN,NaN,NaN,1.856403,0.579261,0.0,2.0,2.0,2.0,55.0



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
company,112593,94.31
agent,16340,13.69
country,488,0.41
hotel,0,0.00
arrival_date_month,0,0.00
arrival_date_week_number,0,0.00
lead_time,0,0.00
is_canceled,0,0.00
stays_in_weekend_nights,0,0.00
stays_in_week_nights,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,31994,26.8



[7] CARDINALITY


,Unique Values
adr,8879
reservation_status_date,926
lead_time,479
company,352
agent,333
country,177
days_in_waiting_list,128
previous_bookings_not_canceled,73
arrival_date_week_number,53
stays_in_week_nights,35



DATASET: CUSTOMERS

[1] SAMPLE DATA (10 ROWS)


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
69432,ca98d5e54ce09ded249a3bf88adda456,c079abb4d146311eb113fffc850247f2,32215,contagem,MG
2225,71cae8ce2f6db9655de76e212d34fe56,9aba983b0a4cf4d7130cf1709070d0f3,6473,barueri,SP
67106,b435b489e4d9921ea0ac305e4e9ae452,662b79d4f9d63b429cf1efcdeacfe6a1,35519,nova serrana,MG
1592,bbe816e2a62258c7b4328b3cecc3caa3,d3a5fa065a68c1f357e6d8f4385938ee,1224,sao paulo,SP
4119,076933dfa56c84230b2965c7cab486e1,ee0fe6783108a6ea35f38c70b6cf08b2,3731,sao paulo,SP
14156,241a20babe281dff9201cd9abc43b96a,2e2fa0fb69bc20c284e174eda80bf5a4,12072,taubate,SP
37182,c1b9ff12f13a4f0d2c4c36b7089eaa93,eef2760c32898a159a068d9acac740cd,31070,belo horizonte,MG
12689,6e38fceafd263d906a4a198e5c5f98a2,49bbdba74838070fe0bdc269cf217cde,39980,cachoeira de pajeu,MG
36258,5ce679079249b5d4d0ce9932aebee695,087329e89b9294b9f586b7d7392537e8,63560,acopiara,CE
61555,15df61aa45e8b81407e1dbcaf8a975fc,db1522683081086b5175654322308a04,86240,sao sebastiao da amoreira,PR



[2] DATA SHAPE


,Rows,Columns
0,99441,5



[3] DATA TYPES


,Column,Data Type
customer_id,customer_id,object
customer_unique_id,customer_unique_id,object
customer_zip_code_prefix,customer_zip_code_prefix,int64
customer_city,customer_city,object
customer_state,customer_state,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
customer_id,99441,99441,274fa6071e5e17fe303b9748641082c8,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_unique_id,99441,96096,8d50f5eadf50201ccdcedfb9e2ac8455,17,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_zip_code_prefix,99441.0,NaN,NaN,NaN,35137.474583,29797.938996,1003.0,11347.0,24416.0,58900.0,99990.0
customer_city,99441,4119,sao paulo,15540,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_state,99441,27,SP,41746,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
customer_id,0,0.0
customer_unique_id,0,0.0
customer_zip_code_prefix,0,0.0
customer_city,0,0.0
customer_state,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
customer_id,99441
customer_unique_id,96096
customer_zip_code_prefix,14994
customer_city,4119
customer_state,27



DATASET: GEOLOCATION

[1] SAMPLE DATA (10 ROWS)


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
940859,90230,-30.008802,-51.202747,porto alegre,RS
407210,20241,-22.925695,-43.190727,rio de janeiro,RJ
654227,38705,-18.595037,-46.496420,patos de minas,MG
514486,28360,-21.132455,-41.690538,bom jesus do itabapoana,RJ
904553,88047,-27.624911,-48.525282,florianopolis,SC
144012,5890,-23.664655,-46.785571,sao paulo,SP
703655,48860,-10.975687,-39.623296,queimadas,BA
301996,13331,-23.073053,-47.226078,indaiatuba,SP
375889,17740,-21.727273,-50.723523,rinopolis,SP
243822,9961,-23.699839,-46.608115,diadema,SP



[2] DATA SHAPE


,Rows,Columns
0,1000163,5



[3] DATA TYPES


,Column,Data Type
geolocation_zip_code_prefix,geolocation_zip_code_prefix,int64
geolocation_lat,geolocation_lat,float64
geolocation_lng,geolocation_lng,float64
geolocation_city,geolocation_city,object
geolocation_state,geolocation_state,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
geolocation_zip_code_prefix,1000163.0,NaN,NaN,NaN,36574.166466,30549.33571,1001.0,11075.0,26530.0,63504.0,99990.0
geolocation_lat,1000163.0,NaN,NaN,NaN,-21.176153,5.715866,-36.605374,-23.603546,-22.919377,-19.97962,45.065933
geolocation_lng,1000163.0,NaN,NaN,NaN,-46.390541,4.269748,-101.466766,-48.573172,-46.637879,-43.767709,121.105394
geolocation_city,1000163,8011,sao paulo,135800,NaN,NaN,NaN,NaN,NaN,NaN,NaN
geolocation_state,1000163,27,SP,404268,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
geolocation_zip_code_prefix,0,0.0
geolocation_lat,0,0.0
geolocation_lng,0,0.0
geolocation_city,0,0.0
geolocation_state,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,261831,26.18



[7] CARDINALITY


,Unique Values
geolocation_lng,717613
geolocation_lat,717360
geolocation_zip_code_prefix,19015
geolocation_city,8011
geolocation_state,27



DATASET: ORDER_ITEMS

[1] SAMPLE DATA (10 ROWS)


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
91802,d0467704fe947b94f1fee589531cb69f,2,3918bb60ff48662029ca9f1e1f7b9842,d57e18d5f73c7ccb7f7339b61166898d,2018-01-11 17:48:10,35.9,15.11
86477,c467181d4844c139feaec6947fd94606,1,71a7800a633691de8ecdd17463335e2e,f80edd2c5aaa505cc4b0a3b219abf4b8,2017-05-07 22:31:45,59.9,24.91
86265,c3f309f7d0ad0cf3b17a58447ed8ade8,1,79fc4c7da6812091c88d89e02e02887c,44073f8b7e41514de3b7815dd0237f4f,2017-11-08 12:55:29,95.0,29.07
48272,6dbeb47d61ce0769596f48e01978dafa,2,e3e6af2b9a779d64b3589a5b56102d16,5343d0649eca2a983820bfe93fc4d17e,2018-03-30 14:35:22,39.0,8.88
6163,0e0441786430b7cfcd0d91f822ea4c99,1,a5ae8400fc9fcacd4af585a57bbf264a,66922902710d126a0e7d26b0e3805106,2018-02-22 16:10:46,105.0,16.49
28851,41a207671f292d00ffea01b171f2a30b,2,f883c296edea388761ae3de2bc4870f3,6560211a19b47992c3666cc44a7e94c0,2017-09-11 09:32:33,45.0,7.78
79329,b47238f009b4f97b2676067d6c0c0386,1,98224bfc1eaadb3a394ec334c60453ff,ce616e1913288884e7742faac9d981db,2018-08-16 10:10:15,4.3,7.39
3315,0771dcbed01969791b77f20d05573b89,1,095cb863f2b04911455ad33f5177062e,2059c39f76271d4ca3f15b5ffaccc8b8,2017-07-06 17:55:20,53.9,13.40
56174,7fc89538fb86a1a09c3e0d809a64f17a,1,d1c427060a0f73f6b889a5c7c61f2ac4,a1043bafd471dff536d0c462352beb48,2018-02-15 12:10:52,149.0,57.27
103767,eba72fc69b4d8bb6893e635895a7bbf5,1,52e9413ed0d3e64b796aac2bc2291d50,7f2617c58d5d06806987308b45654351,2018-05-16 09:15:16,149.9,12.43



[2] DATA SHAPE


,Rows,Columns
0,112650,7



[3] DATA TYPES


,Column,Data Type
order_id,order_id,object
order_item_id,order_item_id,int64
product_id,product_id,object
seller_id,seller_id,object
shipping_limit_date,shipping_limit_date,object
price,price,float64
freight_value,freight_value,float64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
order_id,112650,98666,8272b63d03f5f79c56e9e4120aec44ef,21,NaN,NaN,NaN,NaN,NaN,NaN,NaN
order_item_id,112650.0,NaN,NaN,NaN,1.197834,0.705124,1.0,1.0,1.0,1.0,21.0
product_id,112650,32951,aca2eb7d00ea1a7b8ebd4e68314663af,527,NaN,NaN,NaN,NaN,NaN,NaN,NaN
seller_id,112650,3095,6560211a19b47992c3666cc44a7e94c0,2033,NaN,NaN,NaN,NaN,NaN,NaN,NaN
shipping_limit_date,112650,93318,2017-07-21 18:25:23,21,NaN,NaN,NaN,NaN,NaN,NaN,NaN
price,112650.0,NaN,NaN,NaN,120.653739,183.633928,0.85,39.9,74.99,134.9,6735.0
freight_value,112650.0,NaN,NaN,NaN,19.99032,15.806405,0.0,13.08,16.26,21.15,409.68



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
order_id,0,0.0
order_item_id,0,0.0
product_id,0,0.0
seller_id,0,0.0
shipping_limit_date,0,0.0
price,0,0.0
freight_value,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,98666
shipping_limit_date,93318
product_id,32951
freight_value,6999
price,5968
seller_id,3095
order_item_id,21



DATASET: PAYMENTS

[1] SAMPLE DATA (10 ROWS)


,order_id,payment_sequential,payment_type,payment_installments,payment_value
72329,330e33c1447ae3b325c4d6e64302523a,1,credit_card,8,130.05
19295,cc5a504c7761122bceafca452e63c2de,1,credit_card,6,187.41
40035,cc16ebef1f3a4045ed7084406d057749,1,credit_card,5,94.50
12004,bfdf7d1387c8a71618bc72ce34b1d51f,1,boleto,1,183.47
41884,0d536196878234ebc112ee9c2cd160fd,1,credit_card,1,103.50
58155,0493b5f35d3ee7b1f7e2d98ad782820b,1,credit_card,2,118.66
51283,d314311fb805546efac408edd66415e4,1,credit_card,4,44.11
13885,e8b1509679b2b25a892ef27e75bf6ee2,1,debit_card,1,37.75
85616,19c36691fd116059bf49b7960f1db743,1,boleto,1,88.30
72431,437dbe46155da24f92d6b155899315b8,1,credit_card,1,50.03



[2] DATA SHAPE


,Rows,Columns
0,103886,5



[3] DATA TYPES


,Column,Data Type
order_id,order_id,object
payment_sequential,payment_sequential,int64
payment_type,payment_type,object
payment_installments,payment_installments,int64
payment_value,payment_value,float64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
order_id,103886,99440,fa65dad1b0e818e3ccc5cb0e39231352,29,NaN,NaN,NaN,NaN,NaN,NaN,NaN
payment_sequential,103886.0,NaN,NaN,NaN,1.092679,0.706584,1.0,1.0,1.0,1.0,29.0
payment_type,103886,5,credit_card,76795,NaN,NaN,NaN,NaN,NaN,NaN,NaN
payment_installments,103886.0,NaN,NaN,NaN,2.853349,2.687051,0.0,1.0,1.0,4.0,24.0
payment_value,103886.0,NaN,NaN,NaN,154.10038,217.494064,0.0,56.79,100.0,171.8375,13664.08



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
order_id,0,0.0
payment_sequential,0,0.0
payment_type,0,0.0
payment_installments,0,0.0
payment_value,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,99440
payment_value,29077
payment_sequential,29
payment_installments,24
payment_type,5



DATASET: REVIEWS

[1] SAMPLE DATA (10 ROWS)


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
71545,7454ac458340c307ff75df48cd30cc80,5be7fcdba040de4538c9bd6fc2447973,4,NaN,NaN,2017-09-07 00:00:00,2017-09-07 16:53:15
24126,608f0336f398e892c385aaf0347c4d9f,5504a1a5092b8696eac00711644d4306,5,NaN,NaN,2017-09-09 00:00:00,2017-09-10 05:18:41
60639,1f997992c19d44fe5049ec91f13ccd34,78ac0881f0e98a1afd25f066d5451d70,3,NaN,Produto demorou muito pra chegar.,2018-02-15 00:00:00,2018-02-18 09:19:53
51635,3463f139ec98f65f5f6d1b63eb2e958d,e0d9fab53226e95f004cdbd712a4b06d,5,NaN,excelente serviço\r\n,2017-03-07 00:00:00,2017-03-10 11:54:50
75890,a3d2622729772e64544ec5dbc540c640,c81f74e50f0496fa39716cc77cacd460,5,NaN,NaN,2018-03-14 00:00:00,2018-03-16 02:22:50
47749,96c8bd857db7b00819948ac2dd02650c,f3988a6f6afe000b1617ab4429c5396c,1,NaN,NaN,2017-08-10 00:00:00,2017-08-12 08:26:32
58326,b03cb4cb177d2bc5ee0b381d4f2ef044,d73b26310b53b3e3958a2aac75ad8e6c,5,NaN,NaN,2017-04-18 00:00:00,2017-04-22 02:18:29
18190,1eea5ac71f4d9234cd164b38528dd637,2f7b0d36ffcaf0277583329ded34c97d,5,NaN,NaN,2017-03-14 00:00:00,2017-03-15 13:53:47
87791,a2c4461f4582e5316a9c3851ece9fe3e,3ff313d5a95fbba4b58f90086504022f,5,NaN,NaN,2018-08-04 00:00:00,2018-08-10 13:46:53
80212,24c5acbbd194d012f27357fb64d47df6,3dec8b48a2ca7a9568f4c63cc3fab4f1,5,NaN,NaN,2017-04-07 00:00:00,2017-04-09 22:09:17



[2] DATA SHAPE


,Rows,Columns
0,99224,7



[3] DATA TYPES


,Column,Data Type
review_id,review_id,object
order_id,order_id,object
review_score,review_score,int64
review_comment_title,review_comment_title,object
review_comment_message,review_comment_message,object
review_creation_date,review_creation_date,object
review_answer_timestamp,review_answer_timestamp,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
review_id,99224,98410,08528f70f579f0c830189efc523d2182,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN
order_id,99224,98673,df56136b8031ecd28e200bb18e6ddb2e,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_score,99224.0,NaN,NaN,NaN,4.086421,1.347579,1.0,4.0,5.0,5.0,5.0
review_comment_title,11568,4527,Recomendo,423,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_comment_message,40977,36159,Muito bom,230,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_creation_date,99224,636,2017-12-19 00:00:00,463,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_answer_timestamp,99224,98248,2017-06-15 23:21:05,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
review_comment_title,87656,88.34
review_comment_message,58247,58.70
review_id,0,0.00
review_score,0,0.00
order_id,0,0.00
review_creation_date,0,0.00
review_answer_timestamp,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,98673
review_id,98410
review_answer_timestamp,98248
review_comment_message,36159
review_comment_title,4527
review_creation_date,636
review_score,5



DATASET: ORDERS

[1] SAMPLE DATA (10 ROWS)


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
25545,0bad561c531f95c3d0c3ef2e03ed72de,4012a377e719108fd66522acf06dabd8,delivered,2018-01-19 20:49:57,2018-01-20 17:21:27,2018-01-25 21:24:49,2018-02-01 19:10:29,2018-02-14 00:00:00
36615,f93c490f3040a2c005e2f21d86559dc7,15d1ff96168e59165bb5ccf454b44b85,delivered,2018-04-26 20:15:35,2018-04-26 20:31:52,2018-04-27 14:19:00,2018-05-17 13:20:59,2018-05-24 00:00:00
58558,b259e69b3ba911e0517d03ba13aa5b9e,a16e67c974d246977ab8e252636bc1bd,delivered,2017-12-15 23:07:31,2017-12-18 16:12:52,2017-12-19 18:04:14,2017-12-27 17:04:19,2018-01-16 00:00:00
26369,23c95653bc574efa346f4eee3b29e882,7b78acd8d2e5233b6ea898fd997942af,delivered,2017-11-20 12:47:48,2017-11-20 13:15:41,2017-11-21 22:33:38,2017-12-08 00:21:26,2017-12-08 00:00:00
97969,ad7bbf9e258c09c913e56dcba3433286,3bf822526ed19260365ce453f2271fef,delivered,2017-11-24 10:32:33,2017-11-24 11:51:48,2017-11-29 21:21:14,2017-12-13 19:32:39,2017-12-20 00:00:00
29488,878d7b205a8f5eb546d03baa3404fb4c,d61138db5bdc5abf1775b4420cf6682c,delivered,2017-09-10 13:07:21,2017-09-10 13:23:55,2017-09-11 21:23:06,2017-09-15 21:52:52,2017-10-02 00:00:00
10854,5a528ee290cb0fa6e22acf467ea33c09,f09d58eff599c9908ddf3a4ea15e6348,delivered,2017-03-13 16:15:38,2017-03-13 16:15:38,2017-03-14 06:21:52,2017-03-21 16:15:16,2017-04-04 00:00:00
19951,c9f82228090c2fa7709d20648f66e84f,3476d1493473d35c8248360967722adf,delivered,2017-02-24 15:16:21,2017-02-24 15:30:16,2017-03-03 18:54:05,2017-03-08 07:12:51,2017-03-17 00:00:00
58873,0c7f49fc339a1aa66222dc4af4eeafd7,ea6197a226b1b7b0f681eb146cb70665,delivered,2017-07-04 09:35:30,2017-07-05 03:05:17,2017-07-05 14:14:28,2017-07-17 11:29:30,2017-07-26 00:00:00
20645,40d8fa0e0533d4f1e5e5fe1cbcca3739,e8adb2978a9295aa6e384d5945d4a60d,delivered,2018-03-21 17:16:33,2018-03-23 19:27:37,2018-03-27 17:58:39,2018-04-04 23:34:29,2018-04-10 00:00:00



[2] DATA SHAPE


,Rows,Columns
0,99441,8



[3] DATA TYPES


,Column,Data Type
order_id,order_id,object
customer_id,customer_id,object
order_status,order_status,object
order_purchase_timestamp,order_purchase_timestamp,object
order_approved_at,order_approved_at,object
order_delivered_carrier_date,order_delivered_carrier_date,object
order_delivered_customer_date,order_delivered_customer_date,object
order_estimated_delivery_date,order_estimated_delivery_date,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq
order_id,99441,99441,66dea50a8b16d9b4dee7af250b4be1a5,1
customer_id,99441,99441,edb027a75a1449115f6b43211ae02a24,1
order_status,99441,8,delivered,96478
order_purchase_timestamp,99441,98875,2018-08-02 12:05:26,3
order_approved_at,99281,90733,2018-02-27 04:31:10,9
order_delivered_carrier_date,97658,81018,2018-05-09 15:48:00,47
order_delivered_customer_date,96476,95664,2018-05-08 19:36:48,3
order_estimated_delivery_date,99441,459,2017-12-20 00:00:00,522



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
order_delivered_customer_date,2965,2.98
order_delivered_carrier_date,1783,1.79
order_approved_at,160,0.16
order_id,0,0.00
order_purchase_timestamp,0,0.00
order_status,0,0.00
customer_id,0,0.00
order_estimated_delivery_date,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,99441
customer_id,99441
order_purchase_timestamp,98875
order_delivered_customer_date,95664
order_approved_at,90733
order_delivered_carrier_date,81018
order_estimated_delivery_date,459
order_status,8



DATASET: PRODUCTS

[1] SAMPLE DATA (10 ROWS)


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
24029,1d16c0509e7011656e03c6a582ff6667,beleza_saude,57.0,430.0,1.0,30000.0,55.0,75.0,61.0
26283,01bacce2c74317ed2524bcaabae2e9d4,utilidades_domesticas,40.0,196.0,2.0,1000.0,38.0,39.0,38.0
15277,701cfdaf6cb0e11a4200ec31fab1997e,moveis_decoracao,32.0,420.0,6.0,1067.0,34.0,12.0,29.0
14677,f7c72fa5d77c897b79b13a5f86f27b02,brinquedos,40.0,675.0,5.0,1000.0,36.0,21.0,20.0
10422,aaa8957f9861aff94f589c3a0e1106cd,moveis_decoracao,38.0,303.0,6.0,300.0,16.0,22.0,16.0
12703,fbcd62530875b193a5e21728242d77a4,relogios_presentes,53.0,582.0,2.0,100.0,20.0,7.0,15.0
6094,3ecb2f23ecd8934a20027accead2b058,ferramentas_jardim,39.0,374.0,3.0,700.0,16.0,3.0,12.0
6797,c1124abcc7a081c6f00f97c9135522fb,beleza_saude,53.0,2138.0,3.0,100.0,34.0,2.0,26.0
29154,0185f302dc17abf603edfe7929a0120a,automotivo,59.0,651.0,5.0,10600.0,78.0,4.0,47.0
28902,817e7f7aeb6d2edd3000b35a1de238f0,cama_mesa_banho,60.0,487.0,4.0,1900.0,45.0,10.0,34.0



[2] DATA SHAPE


,Rows,Columns
0,32951,9



[3] DATA TYPES


,Column,Data Type
product_id,product_id,object
product_category_name,product_category_name,object
product_name_lenght,product_name_lenght,float64
product_description_lenght,product_description_lenght,float64
product_photos_qty,product_photos_qty,float64
product_weight_g,product_weight_g,float64
product_length_cm,product_length_cm,float64
product_height_cm,product_height_cm,float64
product_width_cm,product_width_cm,float64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
product_id,32951,32951,106392145fca363410d287a815be6de4,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
product_category_name,32341,73,cama_mesa_banho,3029,NaN,NaN,NaN,NaN,NaN,NaN,NaN
product_name_lenght,32341.0,NaN,NaN,NaN,48.476949,10.245741,5.0,42.0,51.0,57.0,76.0
product_description_lenght,32341.0,NaN,NaN,NaN,771.495285,635.115225,4.0,339.0,595.0,972.0,3992.0
product_photos_qty,32341.0,NaN,NaN,NaN,2.188986,1.736766,1.0,1.0,1.0,3.0,20.0
product_weight_g,32949.0,NaN,NaN,NaN,2276.472488,4282.038731,0.0,300.0,700.0,1900.0,40425.0
product_length_cm,32949.0,NaN,NaN,NaN,30.815078,16.914458,7.0,18.0,25.0,38.0,105.0
product_height_cm,32949.0,NaN,NaN,NaN,16.937661,13.637554,2.0,8.0,13.0,21.0,105.0
product_width_cm,32949.0,NaN,NaN,NaN,23.196728,12.079047,6.0,15.0,20.0,30.0,118.0



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
product_category_name,610,1.85
product_description_lenght,610,1.85
product_name_lenght,610,1.85
product_photos_qty,610,1.85
product_weight_g,2,0.01
product_height_cm,2,0.01
product_length_cm,2,0.01
product_width_cm,2,0.01
product_id,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
product_id,32951
product_description_lenght,2960
product_weight_g,2204
product_height_cm,102
product_length_cm,99
product_width_cm,95
product_category_name,73
product_name_lenght,66
product_photos_qty,19



DATASET: SELLERS

[1] SAMPLE DATA (10 ROWS)


,seller_id,seller_zip_code_prefix,seller_city,seller_state
1324,bb2afe2061d53c6968aa0cffa233a8ec,4307,sao paulo,SP
993,3d4824f20035949c710eaf111f869d39,18015,sorocaba,SP
1235,184725737946a1a938bbbd1a9e978103,9370,maua,SP
1997,2cb4700db635baa1c0d4f90ed27b6669,81070,curitiba,PR
961,c286ec035ee04ba81c97508f0b5df544,81320,curitiba,PR
1772,66fd84ee7528c0c0de85d0f44a73d5ab,98975,campina das missoes,RS
448,75fdd0a732287f9497b179f6bd445d32,9350,maua,SP
1550,191d55f7edb8524d525ed01b20fe32e8,73020,brasilia,DF
1697,b05b432bf68945d4d9783f9e3ca29ee8,16200,birigui,SP
1124,365e4bce19bda0082c3d15d83fdb8bcd,9570,sao caetano do sul,SP



[2] DATA SHAPE


,Rows,Columns
0,3095,4



[3] DATA TYPES


,Column,Data Type
seller_id,seller_id,object
seller_zip_code_prefix,seller_zip_code_prefix,int64
seller_city,seller_city,object
seller_state,seller_state,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
seller_id,3095,3095,9e25199f6ef7e7c347120ff175652c3b,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
seller_zip_code_prefix,3095.0,NaN,NaN,NaN,32291.059451,32713.45383,1001.0,7093.5,14940.0,64552.5,99730.0
seller_city,3095,611,sao paulo,694,NaN,NaN,NaN,NaN,NaN,NaN,NaN
seller_state,3095,23,SP,1849,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
seller_id,0,0.0
seller_zip_code_prefix,0,0.0
seller_city,0,0.0
seller_state,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
seller_id,3095
seller_zip_code_prefix,2246
seller_city,611
seller_state,23



DATASET: CATEGORY_TRANSLATION

[1] SAMPLE DATA (10 ROWS)


,product_category_name,product_category_name_english
30,moveis_escritorio,office_furniture
1,informatica_acessorios,computers_accessories
2,automotivo,auto
36,construcao_ferramentas_ferramentas,costruction_tools_tools
0,beleza_saude,health_beauty
3,cama_mesa_banho,bed_bath_table
51,fashion_esporte,fashion_sport
44,industria_comercio_e_negocios,industry_commerce_and_business
60,la_cuisine,la_cuisine
16,ferramentas_jardim,garden_tools



[2] DATA SHAPE


,Rows,Columns
0,71,2



[3] DATA TYPES


,Column,Data Type
product_category_name,product_category_name,object
product_category_name_english,product_category_name_english,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq
product_category_name,71,71,beleza_saude,1
product_category_name_english,71,71,health_beauty,1



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
product_category_name,0,0.0
product_category_name_english,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
product_category_name,71
product_category_name_english,71


## Why Use .copy() When Creating DataFrames?

Ban đầu dữ liệu được gán như sau:

```python
hotel_df = dfs["hotel_bookings"]
```

Cách gán này không tạo ra một DataFrame mới mà chỉ tạo một tham chiếu (reference) tới cùng một đối tượng DataFrame trong bộ nhớ.

Do đó, mọi thay đổi thực hiện trên `hotel_df` sẽ đồng thời làm thay đổi `dfs["hotel_bookings"]`.

Ví dụ:

```python
hotel_df.drop(columns=["company"], inplace=True)
```

Sau lệnh trên, cột `company` cũng bị xóa khỏi `dfs["hotel_bookings"]`.

Để tránh làm thay đổi dữ liệu gốc (raw data), cần sử dụng phương thức `.copy()`:

```python
hotel_df = dfs["hotel_bookings"].copy()
```

Khi đó, `hotel_df` trở thành một bản sao độc lập của dữ liệu gốc. Mọi thao tác làm sạch dữ liệu chỉ ảnh hưởng tới `hotel_df` mà không làm thay đổi `dfs["hotel_bookings"]`.

Việc sử dụng `.copy()` giúp:

* Bảo toàn dữ liệu gốc để đối chiếu khi cần.
* Đảm bảo kết quả EDA luôn được thực hiện trên dữ liệu nguyên bản.
* Tránh các lỗi phát sinh do vô tình thay đổi dữ liệu trong quá trình tiền xử lý.
* Tách biệt rõ ràng giữa dữ liệu thô (raw data) và dữ liệu đã làm sạch (clean data).


In [3]:
hotel_df = dfs["hotel_bookings"].copy()
customers_df = dfs["customers"].copy()
geolocation_df = dfs["geolocation"].copy()
order_items_df = dfs["order_items"].copy()
payments_df = dfs["payments"].copy()
reviews_df = dfs["reviews"].copy()
orders_df = dfs["orders"].copy()
products_df = dfs["products"].copy()
sellers_df = dfs["sellers"].copy()
category_df = dfs["category_translation"].copy()

## Handling Missing Values in Hotel Booking Dataset

Trong quá trình khám phá dữ liệu, bảng `hotel_bookings` xuất hiện một số cột chứa giá trị thiếu (missing values). Các phương pháp xử lý được lựa chọn dựa trên ý nghĩa nghiệp vụ của từng thuộc tính:

* **company**: Cột này có tỷ lệ thiếu rất cao (94.31%), do đó được loại bỏ khỏi tập dữ liệu vì không mang lại nhiều giá trị cho phân tích và mô hình học máy.
* **agent**: Giá trị thiếu được thay thế bằng `0`, biểu thị khách hàng không đặt phòng thông qua đại lý du lịch.
* **country**: Giá trị thiếu được thay thế bằng `"Unknown"` nhằm giữ lại toàn bộ bản ghi và phản ánh rõ các trường hợp không xác định quốc gia.
* **children**: Giá trị thiếu được thay thế bằng `0`, giả định rằng khách hàng không đi cùng trẻ em trong các trường hợp dữ liệu bị thiếu.

Sau khi xử lý, tập dữ liệu không còn giá trị thiếu ở các cột trên và sẵn sàng cho các bước phân tích và xây dựng mô hình tiếp theo.


In [4]:

hotel_df.drop(columns=["company"], inplace=True)

hotel_df["agent"] = hotel_df["agent"].fillna(0)

hotel_df["country"] = hotel_df["country"].fillna("Unknown")

hotel_df["children"] = hotel_df["children"].fillna(0)

In [5]:
hotel_df.duplicated().sum()

np.int64(32001)

## Handling Missing Values in Review Dataset

Bảng `olist_order_reviews_dataset` chứa hai cột văn bản là `review_comment_title` và `review_comment_message` với tỷ lệ giá trị thiếu khá cao. Tuy nhiên, các giá trị thiếu này không phải là lỗi dữ liệu mà phản ánh trường hợp khách hàng chỉ đánh giá bằng số sao mà không để lại nhận xét bằng văn bản.

Do đó, thay vì xóa các bản ghi hoặc loại bỏ cột, các giá trị thiếu được thay thế bằng:

* `No Title` đối với cột `review_comment_title`
* `No Message` đối với cột `review_comment_message`

Cách xử lý này giúp bảo toàn toàn bộ dữ liệu đánh giá và hỗ trợ cho các bước phân tích văn bản hoặc xây dựng mô hình học máy trong tương lai.


In [6]:
reviews_df["review_comment_title"] = (
    reviews_df["review_comment_title"]
    .fillna("No Title")
)

reviews_df["review_comment_message"] = (
    reviews_df["review_comment_message"]
    .fillna("No Message")
)


## Converting Date Columns to Datetime

Các cột thời gian trong bảng `orders` được chuyển từ kiểu dữ liệu `object` sang `datetime` để hỗ trợ việc phân tích dữ liệu theo thời gian, tính toán thời gian giao hàng và xây dựng đặc trưng cho các mô hình học máy.

Các cột được chuyển đổi bao gồm:

* order_purchase_timestamp
* order_approved_at
* order_delivered_carrier_date
* order_delivered_customer_date
* order_estimated_delivery_date

Trong quá trình chuyển đổi, các giá trị không hợp lệ sẽ được tự động chuyển thành `NaT` bằng tham số `errors='coerce'`.


## Handling Missing Values in Order Date Columns

Trong bảng `orders`, ba cột `order_approved_at`, `order_delivered_carrier_date` và `order_delivered_customer_date` xuất hiện một số giá trị thiếu. Sau khi kiểm tra mối quan hệ với thuộc tính `order_status`, nhận thấy các giá trị thiếu này chủ yếu xuất hiện ở những đơn hàng có trạng thái như `canceled`, `unavailable`, `created` hoặc các đơn hàng chưa hoàn tất quy trình xử lý.

Do đó, các giá trị thiếu này không phải là lỗi dữ liệu mà phản ánh đúng trạng thái thực tế của đơn hàng trong hệ thống:

* `order_approved_at`: Đơn hàng chưa được xác nhận thanh toán hoặc bị hủy trước khi hoàn tất quá trình phê duyệt.
* `order_delivered_carrier_date`: Đơn hàng chưa được bàn giao cho đơn vị vận chuyển.
* `order_delivered_customer_date`: Đơn hàng chưa được giao thành công cho khách hàng.

Việc thay thế các giá trị thiếu bằng một ngày giả định có thể làm sai lệch bản chất nghiệp vụ của dữ liệu và ảnh hưởng đến các phân tích liên quan đến thời gian xử lý, vận chuyển và giao hàng. Vì vậy, các giá trị thiếu trong ba cột này được giữ nguyên để đảm bảo tính chính xác và toàn vẹn của dữ liệu.



In [7]:
date_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

orders_df[date_columns] = orders_df[date_columns].apply(
    pd.to_datetime,
    errors='coerce'
)

display(
    pd.DataFrame({
        "Column": date_columns,
        "Data Type": orders_df[date_columns].dtypes.astype(str)
    })
)

,Column,Data Type
order_purchase_timestamp,order_purchase_timestamp,datetime64[ns]
order_approved_at,order_approved_at,datetime64[ns]
order_delivered_carrier_date,order_delivered_carrier_date,datetime64[ns]
order_delivered_customer_date,order_delivered_customer_date,datetime64[ns]
order_estimated_delivery_date,order_estimated_delivery_date,datetime64[ns]


In [8]:
orders_df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

## Handling Missing Values in Product Dataset

Các giá trị thiếu trong bảng sản phẩm được xử lý dựa trên bản chất của từng thuộc tính.

* `product_category_name` được thay thế bằng `"Unknown Category"` nhằm biểu diễn các sản phẩm chưa xác định được danh mục.
* `product_name_lenght`, `product_description_lenght` và `product_photos_qty` được thay thế bằng giá trị `0`, phản ánh trường hợp sản phẩm không có tên, không có mô tả hoặc không có hình ảnh.
* Các thuộc tính vật lý của sản phẩm gồm `product_weight_g`, `product_length_cm`, `product_height_cm` và `product_width_cm` được thay thế bằng giá trị trung vị (Median). Phương pháp này giúp giảm ảnh hưởng của các giá trị ngoại lai và duy trì phân bố dữ liệu gần với thực tế.

Sau khi xử lý, tập dữ liệu không còn giá trị thiếu và sẵn sàng cho các bước phân tích và xây dựng mô hình học máy.


In [9]:
products_df['product_category_name'] = (
    products_df['product_category_name']
    .fillna('Unknown Category')
)


products_df['product_name_lenght'] = (
    products_df['product_name_lenght']
    .fillna(0)
)

products_df['product_description_lenght'] = (
    products_df['product_description_lenght']
    .fillna(0)
)

products_df['product_photos_qty'] = (
    products_df['product_photos_qty']
    .fillna(0)
)

physical_cols = [
    'product_weight_g',
    'product_length_cm',
    'product_height_cm',
    'product_width_cm'
]

for col in physical_cols:
    products_df[col] = products_df[col].fillna(
        products_df[col].median()
    )

In [10]:
for name, df in {
    "hotel": hotel_df,
    "reviews": reviews_df,
    "orders": orders_df,
    "products": products_df
}.items():

    print(f"\n{name.upper()}")
    print(df.isnull().sum()[df.isnull().sum() > 0])


HOTEL
Series([], dtype: int64)

REVIEWS
Series([], dtype: int64)

ORDERS
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64

PRODUCTS
Series([], dtype: int64)


In [11]:
for name, df in dfs.items():
    print(f"{name}: {df.duplicated().sum()} duplicate rows")

hotel_bookings: 31994 duplicate rows
customers: 0 duplicate rows
geolocation: 261831 duplicate rows
order_items: 0 duplicate rows
payments: 0 duplicate rows
reviews: 0 duplicate rows
orders: 0 duplicate rows
products: 0 duplicate rows
sellers: 0 duplicate rows
category_translation: 0 duplicate rows


## Handling Duplicate Records

Sau khi kiểm tra dữ liệu, chỉ có hai bảng xuất hiện số lượng lớn bản ghi trùng lặp là `hotel_bookings` và `geolocation`.

### Hotel Bookings Dataset

Bảng `hotel_bookings` chứa khoảng 31,994 bản ghi trùng lặp hoàn toàn. Tuy nhiên, các bản ghi này được **giữ nguyên** và không bị loại bỏ.

Lý do là tập dữ liệu không cung cấp khóa định danh duy nhất (Booking ID) cho từng lượt đặt phòng. Do đó, không thể khẳng định các bản ghi giống nhau là dữ liệu lỗi. Trong thực tế, nhiều khách hàng khác nhau có thể thực hiện các đặt phòng với cùng khách sạn, thời gian lưu trú, số lượng khách, giá phòng và trạng thái đặt phòng, dẫn đến các bản ghi có giá trị hoàn toàn giống nhau.

Việc loại bỏ các bản ghi này có thể làm mất các giao dịch hợp lệ, ảnh hưởng đến phân tích kinh doanh và làm thay đổi phân phối dữ liệu khi xây dựng mô hình Machine Learning. Vì vậy, các bản ghi trùng lặp trong bảng `hotel_bookings` được giữ lại.

### Geolocation Dataset

Bảng `geolocation` chứa khoảng 261,831 bản ghi trùng lặp hoàn toàn. Các bản ghi này được **loại bỏ** bằng phương pháp `drop_duplicates()`.

Khác với bảng đặt phòng, bảng `geolocation` là bảng dữ liệu tham chiếu địa lý, bao gồm mã vùng bưu chính, tọa độ địa lý và thông tin vị trí. Nhiều dòng trong bảng chứa chính xác cùng một thông tin về mã vùng, kinh độ, vĩ độ, thành phố và bang.

Các bản ghi trùng lặp này không mang thêm thông tin mới cho quá trình phân tích, đồng thời làm tăng kích thước dữ liệu và thời gian xử lý. Việc loại bỏ chúng giúp giảm dung lượng tập dữ liệu, tăng hiệu quả xử lý và không làm mất bất kỳ thông tin nghiệp vụ quan trọng nào.

### Conclusion

* Giữ nguyên các bản ghi trùng lặp trong bảng `hotel_bookings`.
* Loại bỏ các bản ghi trùng lặp trong bảng `geolocation`.
* Các bảng còn lại không chứa bản ghi trùng lặp nên không cần xử lý thêm.


In [12]:
geolocation_df.head()


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [13]:
duplicate_geo = geolocation_df[
    geolocation_df.duplicated()
]

display(duplicate_geo.head())

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
15,1046,-23.546081,-46.644820,sao paulo,SP
44,1046,-23.546081,-46.644820,sao paulo,SP
65,1046,-23.546081,-46.644820,sao paulo,SP
66,1009,-23.546935,-46.636588,sao paulo,SP
67,1046,-23.546081,-46.644820,sao paulo,SP


In [14]:
geolocation_df.drop_duplicates(inplace=True)

In [15]:
for name, df in {
    "hotel": hotel_df,
    "customers": customers_df,
    "geolocation": geolocation_df,
    "order_items": order_items_df,
    "payments": payments_df,
    "reviews": reviews_df,
    "orders": orders_df,
    "products": products_df,
    "sellers": sellers_df,
    "category": category_df
}.items():

    print(f"\n{name.upper()}")
    print("Shape:", df.shape)
    print("Missing:", df.isnull().sum().sum())
    print("Duplicates:", df.duplicated().sum())


HOTEL
Shape: (119390, 31)
Missing: 0
Duplicates: 32001

CUSTOMERS
Shape: (99441, 5)
Missing: 0
Duplicates: 0

GEOLOCATION
Shape: (738332, 5)
Missing: 0
Duplicates: 0

ORDER_ITEMS
Shape: (112650, 7)
Missing: 0
Duplicates: 0

PAYMENTS
Shape: (103886, 5)
Missing: 0
Duplicates: 0

REVIEWS
Shape: (99224, 7)
Missing: 0
Duplicates: 0

ORDERS
Shape: (99441, 8)
Missing: 4908
Duplicates: 0

PRODUCTS
Shape: (32951, 9)
Missing: 0
Duplicates: 0

SELLERS
Shape: (3095, 4)
Missing: 0
Duplicates: 0

CATEGORY
Shape: (71, 2)
Missing: 0
Duplicates: 0


## Explanation of Duplicate Count Changes After Data Cleaning

Trong quá trình tiền xử lý dữ liệu, số lượng bản ghi trùng lặp (duplicate records) của bảng `hotel_bookings` thay đổi từ **31,994** lên **32,001** bản ghi. Sự thay đổi này không phải do dữ liệu mới được tạo ra mà là kết quả của quá trình chuẩn hóa dữ liệu.

Ban đầu, dữ liệu được tải từ file gốc vào dictionary `dfs`:

```python
dfs["hotel_bookings"]
```

Tại thời điểm này, số lượng bản ghi trùng lặp được xác định là:

```text
31,994 duplicate records
```

Để bảo toàn dữ liệu gốc, một bản sao độc lập được tạo bằng phương thức `.copy()`:

```python
hotel_df = dfs["hotel_bookings"].copy()
```

Khi đó:

```text
dfs["hotel_bookings"]  -> 31,994 duplicates
hotel_df              -> 31,994 duplicates
```

Sau đó, các bước làm sạch dữ liệu được thực hiện trên `hotel_df`:

```python
hotel_df.drop(columns=["company"], inplace=True)

hotel_df["agent"] = hotel_df["agent"].fillna(0)

hotel_df["country"] = hotel_df["country"].fillna("Unknown")

hotel_df["children"] = hotel_df["children"].fillna(0)
```

Các thao tác này làm thay đổi cách Pandas nhận diện các bản ghi trùng lặp. Trước khi làm sạch, một số bản ghi có thể khác nhau do:

* Giá trị NULL trong các cột `agent`, `country` hoặc `children`.
* Giá trị khác nhau trong cột `company`.

Sau khi thay thế các giá trị thiếu và loại bỏ cột `company`, một số bản ghi trước đây được xem là khác nhau trở thành hoàn toàn giống nhau trên tất cả các thuộc tính còn lại. Vì vậy, Pandas nhận diện thêm **7 bản ghi trùng lặp**, làm tổng số duplicate tăng từ:

```text
31,994 → 32,001
```

Lưu ý rằng dữ liệu gốc trong `dfs["hotel_bookings"]` không bị thay đổi nhờ sử dụng `.copy()`. Do đó:

```python
dfs["hotel_bookings"].duplicated().sum()
```

vẫn trả về:

```text
31,994
```

trong khi:

```python
hotel_df.duplicated().sum()
```

trả về:

```text
32,001
```

Đối với bước xử lý duplicate, giá trị cần sử dụng là **32,001**, vì đây là số lượng bản ghi trùng lặp của tập dữ liệu sau khi đã hoàn thành quá trình làm sạch. Đây cũng là tập dữ liệu sẽ được sử dụng cho các bước phân tích dữ liệu, trực quan hóa và xây dựng mô hình Machine Learning tiếp theo.


In [16]:
import os
os.makedirs("../data/clean", exist_ok=True)


hotel_df.to_csv(
    "../data/clean/hotel_bookings_clean.csv",
    index=False
)

customers_df.to_csv(
    "../data/clean/customers_clean.csv",
    index=False
)

geolocation_df.to_csv(
    "../data/clean/geolocation_clean.csv",
    index=False
)

order_items_df.to_csv(
    "../data/clean/order_items_clean.csv",
    index=False
)

payments_df.to_csv(
    "../data/clean/payments_clean.csv",
    index=False
)

reviews_df.to_csv(
    "../data/clean/reviews_clean.csv",
    index=False
)

orders_df.to_csv(
    "../data/clean/orders_clean.csv",
    index=False
)

products_df.to_csv(
    "../data/clean/products_clean.csv",
    index=False
)

sellers_df.to_csv(
    "../data/clean/sellers_clean.csv",
    index=False
)

category_df.to_csv(
    "../data/clean/category_translation_clean.csv",
    index=False
)

print("All cleaned datasets have been saved successfully!")

All cleaned datasets have been saved successfully!


In [21]:


products_df.shape

(32951, 9)